### PR TITLE
Allow applications using quakus-info to contribute data to the /info using CDI

### LINKED ISSUE #37994: Provide a way for applications to contribute to the `/info` enpoint
### Description

As of now, the only way to contribute to the `/info` endpoint offered by the quarkus-info extension is to create an extension that can produce `InfoBuildTimeContributorBuildItem`. 

It woud be great to be able to contribute to the endpoint without having to create an extension.

If the extension scan for application defined `InfoContributor` classes, anyone could contribute to the endpoint, kind of how it's done for smallrye-health.

I can try to provide an MR if you like.

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by jtama**
- Closes #37994

### CODE REVIEW

**Comment by jtama** [comment]
@geoand

**Comment by geoand** [comment]
Thanks, I'll have a look tomorrow

**Comment by geoand** [comment]
So if I understand correctly you want to make all implementations of `InfoContributor` beans automatically?

I think it would be safer to have users opt-in to that by explicitly annotating their implementations with any of the standard CDI annotations.

**Comment by geoand** [comment]
Another issue I see is that classes under `runtime` are considered internal, so we would probably want to move `InfoContributor`

**Comment by jtama** [comment]
> So if I understand correctly you want to make all implementations of `InfoContributor` beans automatically?
> 
> I think it would be safer to have users opt-in to that by explicitly annotating their implementations with any of the standard CDI annotations.

Ok I can remove the scan then, and just let user opt-in.

> Another issue I see is that classes under `runtime` are considered internal, so we would probably want to move `InfoContributor`

Ok do you want me to move it to `io.quarkus.info.spi` ? 

Also even If i did test the extension on a local project I can't find any integration-test for this extension AND

**Review by melloware — APPROVED** [review]

**Comment by geoand** [comment]
> Ok do you want me to move it to io.quarkus.info.spi ?

I would say move them to `io.quarkus.info`  as it is no longer an SPI :)

**Comment by jtama** [comment]
> > Ok do you want me to move it to io.quarkus.info.spi ?
> 
> I would say move them to `io.quarkus.info` as it is no longer an SPI :)
> 
> > Also even If i did test the extension on a local project I can't find any integration-test for this extension
> 
> We have a bunch of tests in [here](https://github.com/quarkusio/quarkus/tree/7db1ebb56b984a9e0d4747ab82cc4cbd7d516e80/extensions/info/deployment/src/test/java/io/quarkus/info/deployment)

Ok in. this case shoulld I completely remove the runtime-spi module ?

**Comment by geoand** [comment]
Actually now that I think about it, let's just leave the class in the module and package that it already resides. 
The package is not great, but it's not bad enough to break extensions that already use it

**Comment by jtama** [comment]
Ok so I have something working with opt-in. But that has a drawback.

As long as now one asks specificly to get those `InfoContributor` instances injected, they won't be created.

Soi for this to work I had to the following to the runtime module

```java
  @ApplicationScoped
public class ExternalContributorReporter {

    @Inject
    Instance<InfoContributor> contributors;


    public void report(Map<String, Object> finalBuildInfo) {
        contributors.forEach(contributor -> finalBuildInfo.put(contributor.name(), contributor.data()));
    }
}
``` 

And use the reporter in the `InfoRecorder` class.

Not sure what we prefere.

**Comment by geoand** [comment]
> As long as now one asks specificly to get those InfoContributor instances injected, they won't be created.

The extension can make the beans of type `InfoContributor` unremovable.
That will ensure that they are always added to the application

File: extensions/info/runtime/src/main/java/io/quarkus/info/runtime/InfoRecorder.java [hunk 1]
```diff
@@ -143,6 +145,10 @@ public InfoHandler(Map<String, Object> buildTimeInfo, List<InfoContributor> know
                 // also, do we want to merge information or simply replace like we are doing here?
                 finalBuildInfo.put(contributor.name(), contributor.data());
             }
+            Arc.container().listAll(InfoContributor.class)
```
Comment (geoand) [inline extensions/info/runtime/src/main/java/io/quarkus/info/runtime/InfoRecorder.java:148]: Just a small nitpick: We generally avoid streams and lambdas in runtime code.

**Comment by jtama** [comment]
I think we have something nice now

**Review by geoand — COMMENTED** [review]
Looks great!

I just added a small comment worth addressing

File: extensions/info/runtime/src/main/java/io/quarkus/info/runtime/InfoRecorder.java (hunk 1 above)
Comment (jtama) [reply extensions/info/runtime/src/main/java/io/quarkus/info/runtime/InfoRecorder.java:148]: Yes I know, it's just really deep into my hands now, so I tend to forget ;)

**Review by geoand — APPROVED** [review]

**Comment by jtama** [comment]
Fixed format.

**Pull request merged**